tools/probes/probes_main.c

```diff
@@ -0,0 +1,288 @@
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define APP_NAME "sof-probes"
+
+#define DATA_READ_LIMIT 1000
+#define FILES_LIMIT	8
+
+struct wave_files {
+	FILE *fd;
+	uint32_t buffer_id;
+	uint32_t size;
+	struct wave header;
+};
+
+enum p_state {
+	READY = 0,
+	SYNC,
+	CHECK
+};
+
+static uint32_t sample_rate[] = {
+	8000, 11025, 12000, 16000, 22050, 24000, 32000, 44100,
+	48000, 64000, 88200, 96000, 128000, 176400, 192000
+};
+
+static void usage(void)
+{
+	fprintf(stdout, "Usage %s <option(s)> <buffer_id/file>\n\n", APP_NAME);
+	fprintf(stdout, "%s:\t -p file\tParse extracted file\n\n", APP_NAME);
+	fprintf(stdout, "%s:\t -h \t\tHelp, usage info\n", APP_NAME);
+	exit(0);
+}
+
+int write_data(char *path, char *data)
+{
+	FILE *fd;
+
+	fd = fopen(path, "w");
+	if (!fd) {
+		fprintf(stderr, "error: unable to open file %s, error %d\n",
+			path, errno);
+		return errno;
+	}
+
+	fprintf(fd, "%s", data);
+	fclose(fd);
+
+	return 0;
+}
+
+int get_buffer_file(struct wave_files *files, uint32_t buffer_id)
+{
+	int i;
+
+	for (i = 0; i < FILES_LIMIT; i++) {
+		if (files[i].buffer_id == buffer_id)
+			return i;
+	}
+	return -1;
+}
+
+int init_wave(struct wave_files *files, uint32_t buffer_id, uint32_t format)
+{
+	char path[64];
+	int i;
+
+	i = get_buffer_file(files, 0);
+	if (i == -1) {
+		fprintf(stderr, "error: too many buffers\n");
+		exit(0);
+	}
+
+	fprintf(stdout, "%s:\t Creating wave file for buffer id: %d\n",
+		APP_NAME, buffer_id);
+
+	sprintf(path, "buffer_%d.wav", buffer_id);
+
+	files[i].fd = fopen(path, "wb");
+	if (!files[i].fd) {
+		fprintf(stderr, "error: unable to create file %s, error %d\n",
+			path, errno);
+		exit(0);
+	}
+
+	files[i].buffer_id = buffer_id;
+
+	files[i].header.riff.chunk_id = HEADER_RIFF;
+	files[i].header.riff.format = HEADER_WAVE;
+	files[i].header.fmt.subchunk_id = HEADER_FMT;
+	files[i].header.fmt.subchunk_size = 16;
+	files[i].header.fmt.audio_format = 1;
+	files[i].header.fmt.num_channels = ((format & PROBE_MASK_NB_CHANNELS) >> PROBE_SHIFT_NB_CHANNELS) + 1;
+	files[i].header.fmt.sample_rate = sample_rate[(format & PROBE_MASK_SAMPLE_RATE) >> PROBE_SHIFT_SAMPLE_RATE];
+	files[i].header.fmt.bits_per_sample = (((format & PROBE_MASK_CONTAINER_SIZE) >> PROBE_SHIFT_CONTAINER_SIZE) + 1) * 8;
+	files[i].header.fmt.byte_rate = files[i].header.fmt.sample_rate *
+					files[i].header.fmt.num_channels *
+					files[i].header.fmt.bits_per_sample / 8;
+	files[i].header.fmt.block_align = files[i].header.fmt.num_channels *
+					  files[i].header.fmt.bits_per_sample / 8;
+	files[i].header.data.subchunk_id = HEADER_DATA;
+
+	fwrite(&files[i].header, sizeof(struct wave), 1, files[i].fd);
+
+	return i;
+}
+
+void finalize_wave_files(struct wave_files *files)
+{
+	uint32_t i, chunk_size;
+
+	for (i = 0; i < FILES_LIMIT; i++) {
+		if (files[i].fd) {
+			files[i].size *= 8;
+			chunk_size = files[i].size + 36;
+
+			fseek(files[i].fd, sizeof(uint32_t), SEEK_SET);
```

Comment:
and here, lots of magic numbers too

Reply:
Found a way to get rid of them but not sure if it is more clearer now, please check